### PR TITLE
BREAKING: Fix ID token validation

### PR DESCRIPTION
**Pull request by joshcanhelp**
### Changes

As a part of our ongoing efforts to improve security in our SDKs, we audited this SDK's capability to accurately and securely validate ID tokens. During this audit, we found an issue for applications that were not configuring the SDK correctly.

Applications that did not indicate an expected ID token algorithm were validating the signature based on the `alg` value in the token, allowing both types to be accepted (when Auth0 would only issue them in one specific algorithm). Auth0 only issues ID tokens based on the algorithm set for the application. The default SDK configuration would allow an application set to `RS256` to accept and successfully verify `HS256` ID tokens. 

**The primary breaking change in this PR only affects applications that are using HS256 ID tokens without setting the config key `id_token_alg` to `'HS256'`.** These applications will fail during the callback phase with a `Algorithm not allowed` exception. Applications that are using RS256 ID tokens, regardless of configuration, will not be affected. 

The only change that needs to be made for applications using HS256 is to pass an `id_token_alg` key to the configuration array set to `"HS256"`, like so:

```php
$auth0 = new Auth0([
  'domain' => 'YOUR_DOMAIN',
  'client_id' => 'YOUR_CLIENT_ID',
  'client_secret' => 'YOUR_CLIENT_SECRET',
  'redirect_uri' => 'https://YOUR_APP/callback',
  'id_token_alg' => 'HS256', // Add this line
]);
```

Applications using RS256 can omit that key, since the default will be set to RS256.

**A secondary breaking change in this PR is for any application using the `JWTVerifier` class.** The verifier class will now use the presence of a `client_secret` to determine the token's algorithm before attempting to decode it. This means that applications using the `JWTVerifier` class directly, passing the config key `supported_algs` incorrectly set to `['HS256', 'RS256']`, passing the config key `client_secret`, and attempting to validate an RS256 token will now fail. HS256 tokens configured this way *should* be changed to only allow HS256 tokens using the `supported_algs` key but will continue to function normally without.

The correct way to configure and validate RS256 ID tokens is to pass an array with a single value `'RS256'` and omit the client secret, like so:

```php
$verifier_config = [
    'valid_audiences' => [ 'YOUR_CLIENT_ID' ],
    'supported_algs' => [ 'RS256' ],
    'authorized_iss' => [ 'https://your-tenant.auth0.com/' ],
];
$jwt_verifier          = new JWTVerifier( $verifier_config );
$decoded_token = $jwt_verifier->verifyAndDecode( $id_token );
```

**Other non-breaking changes in this PR:**

- The Client ID will always be included as a valid audience for ID token validation when using the `Auth0` class.
- The Domain as a URL will always be included as an authorized issuer for ID token validation when using the `Auth0` class.
- The Client Secret will not be passed to the `JWTVerifier` class if an RS256 algorithm is indicated  when using the `Auth0` class.
- Invalid issuer and audience exceptions will not include config or token values in their output

### Testing

- [x] This change adds test coverage
- [x] This change has been tested on PHP 7.1

### CODE REVIEW

File: src/Auth0.php [hunk 1]
```diff
@@ -315,22 +315,28 @@ public function __construct(array $config)
...
             if (! is_array( $config['id_token_aud'] )) {
                 throw new CoreException('Invalid id_token_aud; must be an array of string values');
             }
 
-            $this->idTokenAud = $config['id_token_aud'];
+            $this->idTokenAud = array_filter( array_merge( $this->idTokenAud, $config['id_token_aud'] ) );
         }
 
+        // Make sure the Domain is included as a valid issuer
+        $this->idTokenIss = [ 'https://'.$this->domain.'/' ];
```
Comment (lbalmaceda) [inline src/Auth0.php]: when the user uses a "custom domain", will it end up living in the `domain` variable?
Asking because I don't recall now which issuer auth0 uses in that case.

File: src/Auth0.php [hunk 2]
```diff
@@ -315,22 +315,28 @@ public function __construct(array $config)
             $this->idTokenAlg = $config['id_token_alg'];
```
Comment (lbalmaceda) [inline src/Auth0.php]: won't this override the default value, even when a config value is not set?

File: src/JWTVerifier.php [hunk 3]
```diff
@@ -177,53 +174,35 @@ public function verifyAndDecode($jwt)
...
-
             if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
-                throw new CoreException('We cannot trust on a token issued by `'.$body_decoded->iss.'`');
+                throw new InvalidTokenException('Invalid token issuer');
             }
 
-            $jwks_url                   = $body_decoded->iss.$this->jwks_path;
-            $secret[$head_decoded->kid] = $this->JWKFetcher->requestJwkX5c($jwks_url, $head_decoded->kid);
+            $jwks_url = $this->authorized_iss[0].$this->jwks_path;
+            $secret   = $this->JWKFetcher->requestCompleteJwks($jwks_url);
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: can we rename this to "publicKey" or similar? secret sounds symmetric

File: src/JWTVerifier.php [hunk 4]
```diff
@@ -177,53 +174,35 @@ public function verifyAndDecode($jwt)
...
-            }
-
             if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
-                throw new CoreException('We cannot trust on a token issued by `'.$body_decoded->iss.'`');
+                throw new InvalidTokenException('Invalid token issuer');
             }
 
-            $jwks_url                   = $body_decoded->iss.$this->jwks_path;
-            $secret[$head_decoded->kid] = $this->JWKFetcher->requestJwkX5c($jwks_url, $head_decoded->kid);
+            $jwks_url = $this->authorized_iss[0].$this->jwks_path;
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: what is this line doing?

File: tests/Auth0Test.php [hunk 5]
```diff
@@ -35,6 +35,7 @@ class Auth0Test extends \PHPUnit_Framework_TestCase
         'store' => false,
         'state_handler' => false,
         'scope' => 'openid offline_access',
+        'id_token_alg' => 'HS256',
```
Comment (lbalmaceda) [inline tests/Auth0Test.php]: why is this needed here?

File: src/Auth0.php [hunk 6]
```diff
@@ -315,22 +315,28 @@ public function __construct(array $config)
...
+        $this->idTokenIss = [ 'https://'.$this->domain.'/' ];
+
         // If a token issuer is passed, make sure it's an array.
         if (! empty($config['id_token_iss'])) {
             if (! is_array( $config['id_token_iss'] )) {
                 throw new CoreException('Invalid id_token_iss; must be an array of string values');
             }
 
-            $this->idTokenIss = $config['id_token_iss'];
+            $this->idTokenIss = array_filter( array_merge( $this->idTokenIss, $config['id_token_iss'] ) );
```
[the remark this answers is not in the document]
Comment (gkwang) [reply src/Auth0.php]: @joshcanhelp Interestingly the behavior I'm seeing is that the id token is always the auth0 url (while access token can be either depending which /authorize endpoint it's called; but that's out of scope for this PR).  I'm checking with protocol-sessions if that's the right behavior.

I don't think having an array of (trusted) issuers will be a problem as long as the array is not modified after the initial initialization and not blindly trusting the jwt claim. While unlikely, a client may expect 2 different tenants for authentication, and it might just be simpler to offer 1 sdk setup for 2 different tenants.

[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/Auth0.php]: If I use a custom domain, then that custom domain is used in the ID token:

**Custom domain (ROPG)**

Screenshot 2019-08-29 15 49 38

**Tenant domain (ROPG)**

Screenshot 2019-08-29 15 51 06

Maybe I'm misunderstanding your comment?

Either way, I think we're in agreement that using the token issuer to pull the JWKS if fine if an invalid issuer will eventually fail the process. In other words, that was not a security risk in this case (though it's now pulling from the passed-in value).

File: src/Auth0.php [hunk 7]
```diff
@@ -315,22 +315,28 @@ public function __construct(array $config)
...
+        $this->idTokenAud = [ $this->clientId ];
+
         // If a token audience is passed, make sure it's an array.
         if (! empty($config['id_token_aud'])) {
             if (! is_array( $config['id_token_aud'] )) {
                 throw new CoreException('Invalid id_token_aud; must be an array of string values');
             }
 
-            $this->idTokenAud = $config['id_token_aud'];
+            $this->idTokenAud = array_filter( array_merge( $this->idTokenAud, $config['id_token_aud'] ) );
```
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/Auth0.php]: @joshcanhelp Yes. Audience is always included and is set to the auth0 client id.

File: src/Auth0.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/Auth0.php]: > using the token issuer to pull the JWKS if fine if an invalid issuer will eventually fail the process
No. **Don't trust the token contents until you've first verified the signature** and then the claims. And given that for the signature verification you'd need the key, you can't just take it from the `iss` claim of that still not verified token.

In the scenario of custom domains, normally the SDK needs to know not just the custom domain but also the auth0 one (In case the audience needs to be api v2 or userinfo, for example as those would keep that auth0 host). In that situation I think it should be OK to have an array of issuers. However.. if you know the token will always contain the same value in the `iss` claim, the custom domain value, then there's no need to allow a second one by using an array. 

The less doors we left open, the safer. I'm pinging @gkwang to confirm this via protocols-and-sessions, though you've already tested the results.

File: src/JWTVerifier.php (hunk 3 above)
Comment (lbalmaceda) [reply src/JWTVerifier.php]: My point is that class is called "JWKFetcher" and if its purpose is to call the JWKs endpoint and convert that cert hash into a pubkey, it shouldn't be handling HS secrets, only RS keys.

File: src/Auth0.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/Auth0.php]: > No. Don't trust the token contents until you've first verified the signature and then the claims.

As you can see in this PR, that's not happening. Just making the point that this part of the fix is not addressing a security issue, that's all.

**Comment by joshcanhelp** [comment]
@lbalmaceda - Everything is address here.

File: src/JWTVerifier.php [hunk 8]
```diff
@@ -138,12 +138,16 @@ public function __construct(array $config, JWKFetcher $jwkFetcher = null)
         }
 
         // Only store the client_secret if this is an HS256 token.
+        // This only runs if $this->supported_algs is ['HS256'] or ['HS256', 'RS256'].
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: In what case do we want "supported algs" to allow having two values?

File: src/JWTVerifier.php [hunk 9]
```diff
@@ -177,53 +178,35 @@ public function verifyAndDecode($jwt)
...
-            throw new InvalidTokenException('Token algorithm not found');
-        }
-
-        if (! $this->supportsAlg($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not supported');
-        }
-
         // Validate the token audience, if present.
         if (! empty($body_decoded->aud)) {
             $audience = is_array($body_decoded->aud) ? $body_decoded->aud : [$body_decoded->aud];
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: can we move any claim check below the signature check?

File: src/JWTVerifier.php [hunk 10]
```diff
@@ -177,53 +178,35 @@ public function verifyAndDecode($jwt)
...
+                throw new InvalidTokenException('Invalid token audience');
             }
         }
 
-        if ('HS256' === $head_decoded->alg) {
-            $secret = $this->client_secret;
+        // At this point, the constructor has set the supported algs to ensure only one.
+        if ($this->supportsAlg( 'HS256' )) {
+            $signature_key = $this->client_secret;
         } else {
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: **for the scope of this function**, it could have received an array with more values than just HS256 and RS256. Instead of doing "else" I'd rather you check for explicit "RS256" which is the remaining case, and eventually fail in the else case at the end. Alternatively, at the top of this function write a check that the array doesn't contain anything other than RS256 or HS256.

File: src/JWTVerifier.php [hunk 11]
```diff
@@ -177,53 +178,35 @@ public function verifyAndDecode($jwt)
...
-            if (empty($head_decoded->kid)) {
-                throw new CoreException('Token key ID is missing for RS256 token');
-            }
-
             if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
```
Comment (lbalmaceda) [inline src/JWTVerifier.php]: issuer check should be done for HS256 case as well. And AFTER verifying the signature. (just like I asked for the audience)

File: src/Auth0.php [hunk 12]
```diff
@@ -315,22 +315,19 @@ public function __construct(array $config)
...
-            if (! is_array( $config['id_token_aud'] )) {
-                throw new CoreException('Invalid id_token_aud; must be an array of string values');
-            }
+        // Make sure the Client ID is included as a valid audience
+        $this->idTokenAud = [ $this->clientId ];
 
-            $this->idTokenAud = $config['id_token_aud'];
-        }
+        // Make sure the Domain is included as a valid issuer
+        $this->idTokenIss = [ 'https://'.$this->domain.'/' ];
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/Auth0.php]: Domain throughout the SDK is just that domain, copied straight from the app settings.

File: src/JWTVerifier.php (hunk 8 above)
Comment (joshcanhelp) [reply src/JWTVerifier.php]: We don't. We're forcing one or the other with this change.

File: src/JWTVerifier.php (hunk 9 above)
Comment (joshcanhelp) [reply src/JWTVerifier.php]: It was my impression that we were fixing the big stuff, then addressing the rest in a separate PR. See the WIP branch for the main changes made for this epic.

File: src/JWTVerifier.php (hunk 10 above)
Comment (gkwang) [reply src/JWTVerifier.php]: I agree. Performing explicit checks and failing unexpected values is better.

Comment (joshcanhelp) [reply src/JWTVerifier.php]: The logic in this class is currently a bit hard to follow (mainly to try to avoid more breaking changes) but it already checks that the array only contains HS256 and RS256. So the else here could only mean that it's RS256. The constructor ensures that we only have a single value in there, HS256 if that value is in the supported algs passed (which throws an exception if there is no client secret) or RS256.

File: src/JWTVerifier.php (hunk 11 above)
Comment (joshcanhelp) [reply src/JWTVerifier.php]: Requiring an issuer for HS256 would be another breaking change in this class. Should I proceed?

Comment (joshcanhelp) [reply src/JWTVerifier.php]: I moved the check to after the signature validation and this will pick up checks if valid issuers have been added to the config. I can't force an issuer being added for HS256 tokens to this class without another breaking change (though the helper function that calls this always adds one). Should I add this additional break in this PR or leave it as-is for now?

**Comment by joshcanhelp** [comment]
@gkwang @lbalmaceda - Changes made and a few comments placed. Let me know what else needs to be done here, if anything 👍

**Comment by gkwang** [comment]
After some review this is not a security vulnerability since an adversary cannot bypass authentication/signature without compromising a client's secret or taking over issuer domain; which is essentially the OAuth2/OIDC security assumptions. 

The best practice is however to explicitly pass in the signature algorithm (RS256/HS256) you expect to be issued by auth0 and provide client_secret only when absolutely necessary (eg. expecting HS256 jwts or using private client) to minimize any potential attack surface.

We will be closing this PR in favor of bundling some of the optimizations and changes in the up coming release.
